eureka2-ext/eureka2-eureka1x-rest-api/src/main/java/com/netflix/eureka2/eureka1x/rest/registry/Eureka1xRegistryProxyImpl.java

```diff
@@ -0,0 +1,123 @@
+import rx.Scheduler;
+
+/**
+ * @author Tomasz Bak
+ */
+public class Eureka1xRegistryProxyImpl implements Eureka1xRegistryProxy {
+
+    private static final Logger logger = LoggerFactory.getLogger(Eureka1xRegistryProxyImpl.class);
+
+    private final Map<String, RegistrationHandler> handlers = new HashMap<>();
+    private final EurekaRegistrationClient registrationClient;
+    private final Scheduler scheduler;
+
+    private final LeaseExpiryQueue expiryQueue;
+
+    public Eureka1xRegistryProxyImpl(EurekaRegistrationClient registrationClient, Scheduler scheduler) {
+        this.registrationClient = registrationClient;
+        this.scheduler = scheduler;
+        expiryQueue = new LeaseExpiryQueue(this, scheduler);
+    }
+
+    @Override
+    public void register(InstanceInfo instanceInfo) {
+        synchronized (handlers) {
+            RegistrationHandler handler = handlers.get(instanceInfo.getId());
+            if (handler == null) {
+                handler = new RegistrationHandler(registrationClient, scheduler);
+                handlers.put(instanceInfo.getId(), handler);
+            }
+            handler.register(instanceInfo);
+            expiryQueue.enqueue(handler);
+        }
+    }
+
+    @Override
+    public Result appendMeta(String appName, String instanceId, Map<String, String> meta) {
+        synchronized (handlers) {
+            RegistrationHandler handler = handlers.get(instanceId);
+            if (handler != null) {
+                InstanceInfo v1InstanceInfo = handler.getV1InstanceInfo();
+                if (appName.equalsIgnoreCase(v1InstanceInfo.getAppName())) {
+                    InstanceInfo newInstanceInfo = merge(v1InstanceInfo, meta);
+                    handler.register(newInstanceInfo);
+                    return Result.Ok;
+                }
+                return Result.InvalidArguments;
+            }
+        }
+        return Result.NotFound;
+    }
+
+    @Override
+    public Result unregister(String appName, String instanceId) {
+        synchronized (handlers) {
+            RegistrationHandler handler = handlers.remove(instanceId);
+            if (handler != null) {
+                InstanceInfo v1InstanceInfo = handler.getV1InstanceInfo();
+                if (appName.equalsIgnoreCase(v1InstanceInfo.getAppName())) {
+                    handler.unregister();
+                    return Result.Ok;
+                }
+                return Result.InvalidArguments;
+            }
+        }
+        return Result.NotFound;
+    }
+
+    @Override
+    public void unregister(RegistrationHandler cleanupHandler, long expiryTime) {
+        synchronized (handlers) {
+            String instanceId = cleanupHandler.getV1InstanceInfo().getId();
+            RegistrationHandler activeHandler = handlers.get(instanceId);
+            if (cleanupHandler == activeHandler && activeHandler.getExpiryTime() <= expiryTime) {
+                activeHandler.unregister();
```

Comment:
It really does not matter here as the handler is used internally here and
no client depends on it. Synchronization will force any pending new
registrations/unregistrations to wait.

On Fri, Mar 6, 2015 at 1:39 PM, qiangdavidliu <notifications@github.com>
wrote:

> In
> eureka2-ext/eureka2-eureka1x-rest-api/src/main/java/com/netflix/eureka2/eureka1x/rest/registry/Eureka1xRegistryProxyImpl.java
> <https://github.com/Netflix/eureka/pull/454#discussion_r25980426>:
>
> > +                    handler.unregister();
> > +                    return Result.Ok;
> > +                }
> > +                return Result.InvalidArguments;
> > +            }
> > +        }
> > +        return Result.NotFound;
> > +    }
> > +
> > +    @Override
> > +    public void unregister(RegistrationHandler cleanupHandler, long expiryTime) {
> > +        synchronized (handlers) {
> > +            String instanceId = cleanupHandler.getV1InstanceInfo().getId();
> > +            RegistrationHandler activeHandler = handlers.get(instanceId);
> > +            if (cleanupHandler == activeHandler && activeHandler.getExpiryTime() <= expiryTime) {
> > +                activeHandler.unregister();
>
> remove first from map then call unregister for better synchronisation
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Netflix/eureka/pull/454/files#r25980426>.
>